src/core/ComputeTs.c

```diff
@@ -238,31 +224,90 @@ void _ComputeTs(int snapshot)
 
     // Smooth the density, stars and SFR fields over increasingly larger filtering radii (for evaluating the
     // heating/ionisation integrals)
+    int snapshot_counter_backwards = 1;
+	reion_grids_t* grids = &(run_globals.reion_grids);
+
     for (R_ct = 0; R_ct < TsNumFilterSteps; R_ct++) {
 
       R_values[R_ct] = R;
 
+      if (R_ct == 0) {
+        prev_zpp = zp;
+        prev_R = 0;
+      } else {
+        prev_zpp = zpp_edge[R_ct - 1];
+        prev_R = R_values[R_ct - 1];
+      }
+
+      zpp_edge[R_ct] = prev_zpp - (R - prev_R) * MPC / (drdz((float)prev_zpp)); // cell size
+      zpp = (zpp_edge[R_ct] + prev_zpp) * 0.5; // average redshift value of shell: z'' + 0.5 * dz''
+
+      if (zpp_edge[R_ct] > run_globals.ZZ[snapshot - snapshot_counter_backwards]){
+        if (R_ct==0){
+            weight = dt_dzp * dzp ;//YQ:I would do  dt_dzp using the mid redshift just like dzpp...
+            for (int ii = 0; ii < slab_n_complex; ii++)
+                 grids->sfr[ii] *= weight;
+        }
+        else{
+          weight = dtdz( 0.5 * (run_globals.ZZ[snapshot - snapshot_counter_backwards] + zpp_edge[R_ct-1]));
+          weight *= (run_globals.ZZ[snapshot - snapshot_counter_backwards] - zpp_edge[R_ct-1]);
+          load_reion_sfr_grids(snapshot - snapshot_counter_backwards+1, weight, 1);
+        }
+        while (zpp_edge[R_ct] > run_globals.ZZ[snapshot - snapshot_counter_backwards]){
+          if (zpp_edge[R_ct] < run_globals.ZZ[snapshot - snapshot_counter_backwards - 1]){
+            weight = zpp_edge[R_ct] - run_globals.ZZ[snapshot - snapshot_counter_backwards];
+            weight *= dtdz(0.5*(zpp_edge[R_ct]+run_globals.ZZ[snapshot - snapshot_counter_backwards]));
+          }
+          else{
+            weight = run_globals.ZZ[snapshot - snapshot_counter_backwards - 1] - run_globals.ZZ[snapshot - snapshot_counter_backwards];
+            weight *= dtdz(0.5*(run_globals.ZZ[snapshot - snapshot_counter_backwards - 1] + run_globals.ZZ[snapshot - snapshot_counter_backwards]));
+          }
+          load_reion_sfr_grids(snapshot_counter_backwards, weight, 0);
+          snapshot_counter_backwards+=1;
+        }
+	    if (R_ct==0)
```

Comment:
I think here there is a } missing to close the (zpp_edge[R_ct] > run_globals.ZZ[snapshot - snapshot_counter_backwards]) statement

Reply:
Actually, reading the part below maybe I misunderstood this and this should still be in (zpp_edge[R_ct] > run_globals.ZZ[snapshot - snapshot_counter_backwards]) statement. If that's the case though then I don' understand what is happening here and why this statement can't be included within the lines 246-249

Reply:
there are cases when larger R_ct, i.e. larger radius becomes larger than more than 1 snapshot and, when this happens, we needs to load sfr grids from earlier snapshots.

e.g. this is a run at snapshot 50, and you can see these R_ct actually goes to earlier snapshots:
![image](https://github.com/user-attachments/assets/e6dcb410-d2aa-42bd-a059-2102f162c49a)


Then there are situations at earlier snapshots, when we go back in time too much and pass snapshot zero, there are a couple of lines to deal with that.  .e.g, this is at snapshot 12, and you can see that at R_ct=36, we are already going beyond the earliest snapshot we have 
![Uploading image.png…]()